### PR TITLE
Automatically remove accounts n days after last login

### LINKED ISSUE #295: Remove old accounts
At least `ci-` accounts can be removed. Currently running "Fix file owner in /home/vmail" steps takes a long time because it has to go through a lot of directories.

We also should drop "Fix file owner in /home/vmail" step eventually.

### PR DESCRIPTION
**Pull request by hagenest**
fixes #295

### CODE REVIEW

File: chatmaild/src/chatmaild/ini/chatmail.ini.f [hunk 1]
```diff
@@ -20,6 +20,9 @@
 # days after which mails are unconditionally deleted
 delete_mails_after = 20
 
+# days after which accounts are deleted if there were no login
+delete_accounts_after = 30
```
Comment (missytake) [inline chatmaild/src/chatmaild/ini/chatmail.ini.f:24]: maybe 21 could be the default? This is exactly the GDPR limit iirc. Then we wouldn't have to do any manual action if users want to delete their account, we could just tell them to delete it in their app.

Comment (hagenest) [reply chatmaild/src/chatmaild/ini/chatmail.ini.f:24]: In that case,let's do 20, like with the mails, I'd say

File: chatmaild/src/chatmaild/doveauth.py [hunk 2]
```diff
@@ -106,7 +106,8 @@ def lookup_passdb(db, config: Config, user, cleartext_password):
         if userdata:
             # Update last login time.
             conn.execute(
-                "UPDATE users SET last_login=? WHERE addr=?", (int(time.time()), user)
+                "UPDATE users SET last_login=? WHERE addr=?",
+                (round(int(time.time()) // 86400), user),
```
Comment (hpk42) [inline chatmaild/src/chatmaild/doveauth.py:110]: the `round` seems superflous

File: chatmaild/src/chatmaild/ini/chatmail.ini.f (hunk 1 above)
Comment (hpk42) [reply chatmaild/src/chatmaild/ini/chatmail.ini.f:24]: the GDPR limit is rather 30 days IIRC, so 25 or so should be fine.

File: chatmaild/src/chatmaild/rm_accounts.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    UnixStreamServer,
+)
+
+from .config import Config, read_config
+from .database import Database
+
+
+def remove_users(db: Database, cutoff_date: int):
+    db.connect()
+    try:
```
Comment (hpk42) [inline chatmaild/src/chatmaild/rm_accounts.py:21]: the typical pattern is rather "with db.write_transaction()" like in other places. You don't need explicit committing and finally-clauses then.

File: chatmaild/src/chatmaild/rm_accounts.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        db.commit_changes()
+    finally:
+        db.close()
+
+
+def remove_user_data(db: Database, cutoff_date: int, dir: Path):
+    """Collects all users where last_login < cutoff_date and deletes corresponding directories."""
+    db.connect()
+
+    try:
```
Comment (hpk42) [inline chatmaild/src/chatmaild/rm_accounts.py:33]: same here, use with-statement.

File: chatmaild/src/chatmaild/ini/chatmail.ini.f [hunk 5]
```diff
@@ -20,6 +20,9 @@
 # days after which mails are unconditionally deleted
 delete_mails_after = 20
 
+# days after which accounts are deleted if there were no login
```
Comment (hpk42) [inline chatmaild/src/chatmaild/ini/chatmail.ini.f:23]:
```suggestion
# days after which accounts are deleted if nobody logged in
```

File: chatmaild/src/chatmaild/rm_accounts.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import sys
+import time
+import shutil
+from pathlib import Path
+from socketserver import (
+    StreamRequestHandler,
+    ThreadingMixIn,
+    UnixStreamServer,
+)
+
```
Comment (hpk42) [inline chatmaild/src/chatmaild/rm_accounts.py:14]: remove unused imports.

File: chatmaild/src/chatmaild/rm_accounts.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    """Collects all users where last_login < cutoff_date and deletes corresponding directories."""
+    db.connect()
+
+    try:
+        select_query = "SELECT user FROM users WHERE last_login <?"
+        cursor = db.execute_query(select_query, (cutoff_date,))
+        usernames = cursor.fetchall()
+
+        for username in usernames:
+            user_dir = dir / username[0]
```
Comment (hpk42) [inline chatmaild/src/chatmaild/rm_accounts.py:39]: don't use `dir` -- it's a builtin name and you are shadowing it. 
Also `vmail_basedir` would be more telling.

File: chatmaild/src/chatmaild/rm_accounts.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            if user_dir.exists() and user_dir.is_dir():
+                shutil.rmtree(user_dir)
```
Comment (hpk42) [inline chatmaild/src/chatmaild/rm_accounts.py:41]: probably better with "ignore_errors=True"

**Review by hpk42 — COMMENTED** [review]
apart from the issues mentioned in the comments, also tests are missing.

File: chatmaild/src/chatmaild/ini/chatmail.ini.f (hunk 1 above)
Comment (hagenest) [reply chatmaild/src/chatmaild/ini/chatmail.ini.f]: set it to 25

File: chatmaild/src/chatmaild/doveauth.py (hunk 2 above)
Comment (hagenest) [reply chatmaild/src/chatmaild/doveauth.py:110]: true

**Review by hpk42 — COMMENTED** [review]
PR still  has a  number of problems -- see my comments. 

Note that i am tackling the goal of this PR now myself, because it requires adding tests and touching several code bits to ensure we don't accidently remove users.  Also, we need this PR rather sooner than later because nine needs to be properly cleaned up still.

**Comment by hpk42** [comment]
k, since #350 is merged, this PR is redundant.

**Pull request closed without merging**